File: frontend/src/services/CommercialistaService.js
import BaseService from './BaseService.js';
import Cliente from '../models/Cliente.js';
import Utente from '../models/Utente.js';
import Documento from '../models/Documento.js';
import Pratica from '../models/Pratica.js';

/**
 * Service dell'area commercialista: clienti con relative pratiche e documenti,
 * collaboratori, documenti dello studio e calcolo delle imposte di un cliente.
 */
class CommercialistaService extends BaseService {

  async getClienti() {
    return Cliente.fromList(await this.get('/commercialista/clienti'));
  }

  async getCliente(id) {
    return new Cliente(await this.get(`/commercialista/clienti/${id}`));
  }

  async getClientePratiche(id) {
    return Pratica.fromList(await this.get(`/commercialista/clienti/${id}/pratiche`));
  }

  async getClienteDocumenti(id) {
    return Documento.fromList(await this.get(`/commercialista/clienti/${id}/documenti`));
  }

  async getCollaboratori() {
    return Utente.fromList(await this.get('/commercialista/collaboratori'));
  }

  async getDocumenti() {
    return Documento.fromList(await this.get('/commercialista/documenti'));
  }

  // Ritorna l'importo (Double) calcolato dal backend
  calcolaImposte(clienteId) {
    return this.get(`/commercialista/imposte/${clienteId}`);
  }
}

export default new CommercialistaService();

File: backend/src/main/java/com/tesi/gestionalec/service/interfaces/CommercialistaService.java
package com.tesi.gestionalec.service.interfaces;

import com.tesi.gestionalec.model.Cliente;
import com.tesi.gestionalec.model.Collaboratore;
import com.tesi.gestionalec.model.Documento;
import com.tesi.gestionalec.model.Pratica;
import java.util.List;

/**
 * Contratto specifico per il commercialista: gestione delle pratiche di studio,
 * assegnazioni ai collaboratori, calcolo delle imposte e accesso a clienti,
 * collaboratori e documenti, in aggiunta alle funzionalità comuni ereditate.
 */
public interface CommercialistaService extends UtenteService {
    List<Pratica> trovaTutteLePratiche();

    void assegnaCollaboratore(Long praticaId, Long collaboratoreId);

    void avanzaStatoPratica(Long praticaId);

    double calcolaImposteCliente(Long clienteId);

    /** Lista di tutti i clienti del sistema (accessibile al commercialista). */
    List<Cliente> trovaTuttiClienti();

    /** Clienti associati a uno specifico commercialista. */
    List<Cliente> trovaClientiDelCommercialista(Long commercialistaId);

    /** Collaboratori associati al commercialista (inviti ACCEPTED). */
    List<Collaboratore> trovaMieiCollaboratori(Long commercialistaId);

    /**
     * Documenti dello studio del commercialista specificato (non quelli di altri
     * studi).
     */
    List<Documento> trovaDocumentiStudio(Long commercialistaId);

    /**
     * Verifica che il cliente con clienteId appartenga al commercialista con
     * commercialistaId.
     * Lancia ForbiddenOperationException se il cliente appartiene a un altro
     * commercialista.
     * Lancia ResourceNotFoundException se il cliente non esiste.
     */
    void verificaAppartenenzaCliente(Long clienteId, Long commercialistaId);
}
